### PR TITLE
Mixer

### PR DESCRIPTION
**Pull request by superjax**

### CODE REVIEW

File: src/mixer.c [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+  if (value > 1000)
+  {
+    value = 1000;
+  }
+  else if (value < 0)
+  {
+    value = 0;
+  }
+  pwmWriteMotor(index, value+2000);
```
Comment (dpkoch) [inline src/mixer.c]: Should this be value+1000 so that the resulting PWM is in the 1000us to 2000us range?

File: src/mixer.c [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  else if (value < -500)
+  {
+    value = -500;
+  }
+  pwmWriteMotor(index, 1500+value);
+}
+
+void mix_output()
+{
+  printf("COMMANDS: F = %d, x = %d, y = %d, z = %d\n", _command.F, _command.x, _command.y, _command.z);
```
Comment (dpkoch) [inline src/mixer.c]: I think we probably want to remove printf statements if we're done with using them for debugging?

File: src/mixer.c [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    if (mixer_to_use.output_type[i] != NONE)
+    {
+      // Matrix multiply (in so many words) -- done in integer, hence the /1000 at the end
+      _outputs[i] = (_command.F*mixer_to_use.F[i] + _command.x*mixer_to_use.x[i] +
+                    _command.y*mixer_to_use.y[i] + _command.z*mixer_to_use.z[i])/1000;
+      if (_outputs[i] > 1000 && _outputs[i] > max_output)
+      {
+        max_output = _outputs[i];
+      }
+      printf("mixer col %d: %d %d %d %d\t", i, mixer_to_use.F[i], mixer_to_use.x[i], mixer_to_use.y[i], mixer_to_use.z[i]);
```
Comment (dpkoch) [inline src/mixer.c]: printf

File: src/mixer.c [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+void init_mixing()
+{
+  // We need a better way to choosing the mixer
+  mixer_to_use = *array_of_mixers[QUADCOPTER_X];
+
+  for(int8_t i=0; i<8; i++)
+  {
+    _outputs[i] = 0;
+    _GPIO_outputs[i] = 0;
+    _GPIO_output_type[i] = 0;
```
Comment (dpkoch) [inline src/mixer.c]: Use `NONE` from the `output_type_t` enum here instead of 0?

File: doc/mixer.tex [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+\end{equation}
+
+Where, to avoid doing floating point calculations, it defined with a nominal value of 1000.  After multiplication, the result needs to be divided by 1000 to remain in the same units as the command.  The remaining outputs are made available to modification by the GPIO inputs to the FC via MAVlink.  It is assumed that PID gains in higher-level functions perform the scaling between the different channels to acheive the desired output.
+
+In addition, each mixer contains information about what kind of output each output channel is.  For example, in the case of a fixed-wing.  The output of the mixer for the first channel is attached to a motor while the other three are attached to servos.  The output of a mixer to a servo must lie between -500 and 500, after which it is shifted up to the center of the servo output.  Motor commands, on the other hand, are restricted to lie between 0 and 1000, and mapped to the full PWM range for standard ESC operation.
+
+
+\begin{figure}[h]
+	\centering
+	\includegraphics[scale=0.3]{./fig/motor_layout.jpg}
```
Comment (dpkoch) [inline doc/mixer.tex]: This image didn't get committed

File: doc/mixer.tex [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
...he maximum roll command would actually be interpreted as a half-roll command as the difference between the motors on opposite sides of the MAV would be halved.  To increase controllability in these sorts of situations, we take the motor control value from the mixer, before it is shifted up 1000$\upmu}$s to the standard PWM range, and find the scaling factor which would reduce the maximum motor command to 1000$\upmu}$s.  We then apply this scaling factor to all other motors, and then shift them up to the standard PWM range of 1000$\upmu}$s to 2000$\upmu}$s.  This does not completely solve the problem, because the maximum roll command is still reduced somewhat, but it trades off maximum thrust for some additional controllability.  The problem is really caused by a limit in how much acceleration is really possible in the MAV, not so much the way in which these commands are interpreted.  In the case of a simultaneous maximum roll and maximum throttle, two of the motors will receive a command of 2000$\upmu}$s, while the other two will receive a command of 1333$\upmu}$s, as opposed to the 1500$\upmu}$s without the saturation, and maintains controllability even at high levels of throttle.
```
Comment (dpkoch) [inline doc/mixer.tex]: pdflatex is unhappy with something about the math in line. My guess is that we need to get rid of the `}` after `\upmu` (or add the missing `{` instead if that's necessary)

File: src/mixer.c [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+      // Incorporate GPIO on not already reserved outputs
+      _outputs[i] = _GPIO_outputs[i];
+      output_type = _GPIO_output_type[i];
+    }
+
+    // Write output to motors
+    if (output_type == S)
+    {
+      write_servo(i, _outputs[i]);
+      printf("writing SERVO %d, %d\n", i+1, _outputs[i]);
```
Comment (dpkoch) [inline src/mixer.c]: printf

File: src/mixer.c [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    }
+    else if (output_type == M)
+    {
+      write_motor(i, _outputs[i]);
+      printf("writing MOTOR %d: %d\n", i+1, _outputs[i]);
```
Comment (dpkoch) [inline src/mixer.c]: printf

**Comment by dpkoch** [comment]
I think it's looking pretty good! Just a few cleanup things, but awesome work! It also may not hurt to run the code style script, but overall I think it looks really good.

File: src/mixer.c (hunk 1 above)
Comment (superjax) [inline src/mixer.c]: I'm sorry, you're right

**Comment by superjax** [comment]
I don't have the astyle thing.  What's that?

**Comment by superjax** [comment]
nvm, sudo-apt-get install astyle

File: src/estimator.c [hunk 9]
```diff
@@ -1,5 +1,3 @@
-#include <math.h>
```
Comment (dpkoch) [inline src/estimator.c]: Don't we need this for `abs` and `sign`?

**Comment by dpkoch** [comment]
Sorry for being so picky. That's my only question; other than that I think it looks great and am good to merge!

**Comment by superjax** [comment]
No, I have my own implementation of sign in turbotrig,  I don't know about abs.  I know that it builds though.

**Comment by dpkoch** [comment]
Cool. Looks like `abs` lives in `stdlib.h`; I've added it in. I guess it builds because that's been included up the line somewhere; I just like to have all the required headers in the same file for completeness so weird build errors don't pop up later if the include order changes. Also, turns out `sign` doesn't exist in the standard libraries. Who knew.

**Pull request merged**